sgkit/io/vcfzarr_reader.py

```diff
@@ -0,0 +1,109 @@
+from typing import Any, List, Tuple
+
+import dask.array as da
+import numpy as np
+import xarray as xr
+import zarr
+
+from ..api import DIM_VARIANT, create_genotype_call_dataset
+from ..typing import ArrayLike, PathType
+
+
+def encode_array(x: ArrayLike) -> Tuple[ArrayLike, List[str]]:
+    """Encode array values as integers indexing unique values
+
+    The codes created for each unique element in the array correspond
+    to order of appearance, not the natural sort order for the array
+    dtype.
+
+    Examples
+    --------
+
+    >>> encode_array(['c', 'a', 'a', 'b'])
+    (array([0, 1, 1, 2]), array(['c', 'a', 'b'], dtype='<U1'))
+
+    Parameters
+    ----------
+    x : (M,) array-like
+        Array of elements to encode of any type
+
+    Returns
+    -------
+    indexes : (M,) ndarray
+        Encoded values as integer indices
+    values : ndarray
+        Unique values in original array in order of appearance
+    """
+    # argsort not implemented in dask: https://github.com/dask/dask/issues/4368
+    names, index, inverse = np.unique(x, return_index=True, return_inverse=True)
+    index = np.argsort(index)
+    rank = np.empty_like(index)
+    rank[index] = np.arange(len(index))
+    return rank[inverse], names[index]
+
+
+def read_vcfzarr(path: PathType) -> xr.Dataset:
+    """Read a VCF Zarr file.
+
+    Loads VCF variant, sample, and genotype data as Dask arrays within a Dataset
+    from a Zarr file created using scikit-allel's `vcf_to_zarr` function.
+
+    Since `vcf_to_zarr` does not preserve phasing information, there is no
+    `call/genotype_phased` variable in the resulting dataset.
+
+    Parameters
+    ----------
+    path : PathType
+        Path to the Zarr file.
+
+    Returns
+    -------
+    xr.Dataset
+        The dataset of genotype calls, created using `create_genotype_call_dataset`.
+    """
+
+    vcfzarr = zarr.open_group(str(path), mode="r")
+
+    # Index the contig names
+    variants_chrom = da.from_zarr(vcfzarr["variants/CHROM"]).astype(str)
+    variant_contig, variant_contig_names = encode_array(variants_chrom.compute())
+    variant_contig = variant_contig.astype("int16")
+    variant_contig_names = list(variant_contig_names)
+
+    # For variant alleles, combine REF and ALT into a single array
+    # and calculate the number of alleles so we can set the dtype correctly
+    variants_ref = da.from_zarr(vcfzarr["variants/REF"])
+    variants_alt = da.from_zarr(vcfzarr["variants/ALT"])
+
+    def max_str_len(arr: ArrayLike) -> Any:
+        return arr.map_blocks(
+            lambda s: np.char.str_len(s.astype(str)), dtype=np.int8
+        ).max()
+
+    max_allele_length = max(
+        da.compute(max_str_len(variants_ref), max_str_len(variants_alt))
+    )
+    variants_ref_alt = da.concatenate(
+        [variants_ref.reshape(-1, 1), variants_alt], axis=1
```

Comment:
Is `vcfzarr["variants/ALT"]` guaranteed to be 2D?  I'm wondering if a `variants_alt.reshape(-1, 1)` also makes sense here to guard against the case where it can be 1D like `variants_ref`.

Reply:
I think it can be 1D also.

Reply:
I've now handled this case.